### PR TITLE
Add inhibitor for invalid fstab entries on pseudo-filesystem mountpoints

### PR DESCRIPTION
**Pull request by prjagtap-rh**
Detect /etc/fstab entries that incorrectly define pseudo-filesystem
mountpoints (/proc, /sys, /dev/shm, /run, etc.) with block devices or
wrong filesystem types. Such configurations have been invalid since
RHEL 7 but are typically ignored during normal boot. During the upgrade
the system applies these entries as configured, leading to failures.
    
Use an allowlist of valid virtual filesystem sources and types so
entries referencing UUID=, LABEL=, PARTUUID=, or /dev/* paths are
correctly detected and inhibited.
    
JIRA: RHEL-17842

### CODE REVIEW

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/tests/test_checkfstabapifsoverride.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        ([_SYS_LVM], True),
+        ([_SYS_LVM_SYSFS], True),
+        ([_SYS_UUID], True),
+        ([_SYS_UUID_SYSFS], True),
+        ([_SHM_LABEL], True),
+        ([_RUN_PARTUUID], True),
+        ([_SHM_TMPFS, _PROC_PROC, _SYS_SYSFS], False),
+        ([_PROC_NONE], False),
+        ([_HOME_EXT4, _ROOT_XFS], False),
+        ([], False),
```
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/checkfstabapifsoverride/tests/test_checkfstabapifsoverride.py:83]: I would add a bunch more tests for cases we don't want to inhibit, e.g. normal mount (not kernel api fs) with LABEL, UUID, etc.

Also maybe add some cases bind mounts? e.g. the mount opts contain bind, or the fs_vfstype is mount.

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/actor.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+from leapp.tags import ChecksPhaseTag, IPUWorkflowTag
+
+
+class CheckFstabApiFsOverride(Actor):
+    """
+    Inhibit upgrade if /etc/fstab overrides an API filesystem mountpoint with persistent storage.
+
+    Paths like /proc, /sys, /dev/shm must remain kernel-managed. Overriding
+    them with block devices (via /dev/*, UUID=, LABEL=, etc.) causes the
+    reboot phase to fail.
```
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/checkfstabapifsoverride/actor.py]: Afaik some of the mountpoint we have there are no really API filesystems. E.g. `/run` is populated by userspace software, like systemd.

Also, I wouldn't outright say that they cause the reboot to fail, unless we are sure all of them do, otherwise it's misleading. So I would write that they *might* cause the reboot to fail.

Comment (matejmatuska) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/actor.py]: And also, the case in the ticket doesn't cause a fail on a normal reboot, it just fails and crashes leapp during after the reboot during an upgrade.

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    if not bad_entries:
+        return
+
+    entries_text = ', '.join(f'{e.fs_file} ({e.fs_vfstype})' for e in bad_entries)
+
+    reporting.create_report([
+        reporting.Title(
+            'Persistent filesystem mounted over API filesystem mountpoint. '
+            "Upgrade can't proceed"
```
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: Don't put phrases like "Upgrade can't proceed" into the report titles. I know there might be a few which do have that, but  in general we don't do it.

Comment (matejmatuska) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: Also, as I mentioned in other comment, I don't think its 100% correct to call the filesystem kernel APIs if some of them really aren't, e.g. /run.

I think maybe we could just call them pseudo-filesystems. Let's see what others think.

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        ),
+        reporting.Summary(
+            'Detected /etc/fstab entries that incorrectly override kernel-managed API '
+            'filesystem mountpoints. This can be caused by mounting a block device '
+            '(via /dev/*, UUID=, LABEL=, etc.) or using a wrong filesystem type at '
+            'these paths. This will cause the upgrade to fail during the reboot phase.\n'
+            'Offending entries: {}'.format(entries_text)
```
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: Again with the kernel API, but unsure about that. What I am sure is that the part about the upgrade failing is not needed.
```suggestion
            'Detected /etc/fstab entries that incorrectly override pseudo '
            'filesystem mountpoints. This can be caused by mounting a block device '
            '(via /dev/*, UUID=, LABEL=, etc.) or using a wrong filesystem type at '
            'these paths.'
            'Offending entries: {}'.format(entries_text)
```

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        ),
+        reporting.Remediation(
+            hint=(
+                'Remove the offending entries from /etc/fstab. The kernel and systemd '
+                'mount these paths automatically at boot. If an entry is required for '
```
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: Are you 100% sure this is true for all of them?

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                "compliance, ensure both the source and type are the virtual filesystem "
+                "(e.g. 'tmpfs /dev/shm tmpfs defaults 0 0'), not a block device.\n"
+                'After editing, run "mount -a" to verify, then re-run leapp preupgrade.'
```
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: definitely don't mention stuff like "then re-run leapp preupgrade." it is kind of expected about most of the reports, even more so if they are an inhibitor. It might even be misleading for the user, because they might have run leapp upgrade.
```suggestion
                'After editing, run "mount -a" to verify.'
```

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            )
+        ),
+        reporting.RelatedResource('file', '/etc/fstab'),
+        reporting.Severity(reporting.Severity.HIGH),
+        reporting.Groups([reporting.Groups.FILESYSTEM]),
+        reporting.Groups([reporting.Groups.INHIBITOR]),
```
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: Not a blocker, but you can combine these into a single line:


```suggestion
        reporting.Groups([reporting.Groups.FILESYSTEM, reporting.Groups.INHIBITOR]),
```

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        if not expected:
+            continue
+
+        if entry.fs_vfstype not in expected or entry.fs_spec not in (expected | {'none'}):
+            bad_entries.append(entry)
+
+    if not bad_entries:
+        return
+
+    entries_text = ', '.join(f'{e.fs_file} ({e.fs_vfstype})' for e in bad_entries)
```
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: We now have a `format_list` function to format list of items in a report in a standard way. It was introduced in https://github.com/oamg/leapp/pull/913

**Review by matejmatuska — DISMISSED** [review]
The logic looks good to me, the changes are mostly around wording.
**Don't act on the review yet**, let's wait for opinions from others, @pirat89 can you take a look? Also please go throught the logic, I am not that familiar with possible setups and edge cases. 

I also didn't test it yet, will do now.

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/actor.py (hunk 2 above)
Comment (pirat89) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/actor.py]: It's important to understand that problem is not just that upgrade fails, but that such a definition in FSTAB is invalid basically since RHEL 7 and only reason why it "works" is that such an invalid configurations are usually ignored. The problem it creates during the upgrade is that we tell the system to apply this mess actually as users configured. So we want to be sure that users get rid of it completely.

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py (hunk 4 above)
Comment (pirat89) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: what about to call them just `API filesystem mountpoints` or `system managed API filesystem mountpoints`?  Maybe it would be better if the summary explains better about what we speak, so the term does not have to have the mandatory role here. We can say that these mountpoints are managed by kernel and systemd and they have specific requirements that must be fullfilled.

Pseudo filesystems sounds weird, but I found that the term is used in docs so it seems ok.

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py (hunk 5 above)
Comment (pirat89) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: I suggest to rephrase the hint, saying that we suggest to remove or correct problematic entries

* I would not use word like `offending` as applications do not feel to be offended :-)  
* entries are expected nowadays to be listed if users want to set a specific options to these mounpoints; e.g. limit the size, etc. - e.g. to comply with a certification.

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py (hunk 3 above)
Comment (pirat89) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: also note the title is incorrect as these filesystems are not mounted based on these invalid fstab entries - that's why the system can boot. they are usually ignored during standard system boot. I am more inclining to sth like:
* Detected invalid entries in /etc/fstab for API mountpoints
* Detected invalid entries in /etc/fstab for pseudo-filesystems
* ...

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py (hunk 4 above)
Comment (pirat89) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: I am missing here also explanation why this problem exists. E.g. that such configurations were  common on old systems before the Systemd era, as this is typical mistake made by admins used to RHEL 6 and older systems.

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py (hunk 5 above)
Comment (pirat89) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: also note that information like `The kernel and systemd mount these paths automatically at boot.` is not hint but summary. this is remediation instruction, not explanation of the problem. explanation of the problem is expected to be part of the summary.

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py (hunk 3 above)
Comment (prjagtap-rh) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: @matejmatuska Yes you are right I referred `checkcifs.py` for title. Hence added cant upgrade. But is now corrected.

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py (hunk 8 above)
Comment (prjagtap-rh) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: @matejmatuska thanks for this its of great help will fix it.

File: repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+        if entry.fs_vfstype not in expected or entry.fs_spec not in (expected | {'none'}):
+            bad_entries.append(entry)
+
+    if not bad_entries:
+        return
+
+    entries_text = format_list(
+        ['{} ({})'.format(e.fs_file, e.fs_vfstype) for e in bad_entries],
+        callback_sort=None,
```
Comment (pirat89) [inline repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: @prjagtap-rh  why? what is the problem with the sorted content? unless there is a good reason for it, the content should be sorted by default - it's preferred way.

also, I think that fs_file should be enough to identify the entry. What additional information user get from the fs_vfstype? from my pov it's not saying what's expected, not whether it's actually wrong. I think that we can stay here with fs_file.

Comment (prjagtap-rh) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: My take on not sorting it so that customer would be able to find it easily. 
If sorted and there are multiple entries then it would be confusing.

Also reason I added filesystem to catch scenarios like below where both entries are wrong. [Its less likely case to have multiple entries]
```
UID=023e51f5-... /sys xfs defaults 0 0        
UUID=023e51f5-... /sys sysfs defaults 0 0

```
In such scenarios if filesystem is not added it would be confusing.

Here I tested it on test system by creating below entries in fstab[Considering worst case fstab file]
```
# Bad entries
/dev/mapper/vg00-shm /dev/shm xfs defaults 0 0
LABEL=mydata /dev/shm tmpfs defaults 0 0
/dev/mapper/vg00-shm /dev/shm/ xfs defaults 0 0
/dev/sda5 /proc ext4 defaults 0 0
/dev/mapper/vg00-sys /sys xfs defaults 0 0
/dev/mapper/vg00-sys /sys sysfs defaults 0 0
UUID=023e51f5-... /sys xfs defaults 0 0
UUID=023e51f5-... /sys sysfs defaults 0 0
PARTUUID=abcd-1234 /run ext4 defaults 0 0
UUID=aabbccdd-... /run ext4 defaults 0 0
UUID=12345678-... /dev/shm xfs defaults 0 0
/some/dir /proc none bind 0 0
```

# correct entries
tmpfs /dev/shm tmpfs defaults 0 0
proc /proc proc defaults 0 0
sysfs /sys sysfs defaults 0 0
none /proc proc defaults 0 0
tmpfs /run tmpfs defaults 0 0
/dev/mapper/fedora-home /home ext4 defaults 0 0
UUID=aabbccdd-... /var xfs defaults 0 0
LABEL=optdata /opt ext4 defaults 0 0
/exports/home /home none bind 0 0
/data/shared /mnt/shared none bind 0 0

This is how it reports
```
Problematic entries:  
   - /dev/shm (xfs) 
   - /dev/shm (tmpfs) 
   - /dev/shm/ (xfs) 
   - /proc (ext4) 
   - /sys (xfs) 
   - /sys (sysfs) 
   - /sys (xfs) 
   - /sys (sysfs) 
   - /run (ext4) 
   - /run (ext4) 
   - /dev/shm (xfs) 
   - /proc (none)
```

Comment (prjagtap-rh) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: As discussed on slack I agree this would be a corner case with the worst possible fstab.
Will `sort` and list `fs_file`

Comment (prjagtap-rh) [reply repos/system_upgrade/common/actors/checkfstabapifsoverride/libraries/checkfstabapifsoverride.py]: @pirat89 Fixed

**Comment by matejmatuska** [comment]
The unit tests actions say that your acount is suspended, I think it might be caused by GH Actions issue, from https://www.githubstatus.com/incidents/gnftqj9htp0g:
> We have identified the cause of the authentication issues affecting GitHub Actions and are actively working on mitigation

edit: seems like there are other issue in the other actions as well

**Review by pirat89 — APPROVED** [review]
lgtm and works as expected.
